File: tools/map_convert.pl
#!/usr/bin/perl -w
#script to convert from single char maps to multiple character maps

sub printUsage{
    print "map_convert.pl terrain.cfg map_file.cfg [new_map_file.cfg]\n";
}

sub get_adjacent{
    #returns string of original location+adjacent locations on hex 1-char map
    $x=shift(@_);
    $y=shift(@_);
    local @map=@_;
    foreach(@map){
	chomp;
    }

#    $orig=substr($map[$y],$x,1);
    $odd=($x) % 2;
#    print "$orig($x,$y) : $odd";
    $adj=substr($map[$y],$x,1);
    if($x>0){
	$adj.=substr($map[$y],$x-1,1);
#	print "\tW";
    }
    if($x<length($map[$y])){
	$adj.=substr($map[$y],$x+1,1);
#	print "\tE";
    }
    if($y>0){
	$adj.=substr($map[$y-1],$x,1);
#	print "\tN";
    }
    if($y<$#_){
	$adj.=substr($map[$y+1],$x,1);
#	print "\tS";
    }

    if($x>0 && $y>0 && !$odd){
	$adj.=substr($map[$y-1],$x-1,1);
#	print "\tNW";
    }
    if($x<length($map[$y]) && $y>0 && !$odd){
	$adj.=substr($map[$y-1],$x+1,1);
#	print "\tNE";
    }
    if($x>0 && $y<$#_ && $odd){
	$adj.=substr($map[$y+1],$x-1,1);
#	print "\tSW";
    }
    if($x<length($map[$y]) && $y<$#_ && $odd){
	$adj.=substr($map[$y+1],$x+1,1);
#	print "\tSE";
    }

#    print "\n";
    return($adj);
}

if($#ARGV <1 ||$#ARGV>2){
    printUsage();
    exit;
}

$terrain_file=shift(@ARGV);
$map_file=shift(@ARGV);
if($#ARGV < 0){
    $new_map_file=$map_file;
    $backup=$map_file.".bak";
    system("cp $map_file $backup") && die "could not create backup file: $backup\n";
}else{
    $new_map_file=shift(@ARGV);
    if(-e "$new_map_file"){ die "New map file already exists: $new_map_file\n";}
}

if(! -r "$terrain_file"){ die "can not read terrain file: $terrain_file\n";}
if(! -r "$map_file"){ die "can not read map file: $map_file\n";}

$conversion{1}='1 _K';
$conversion{2}='2 _K';
$conversion{3}='3 _K';
$conversion{4}='4 _K';
$conversion{5}='5 _K';
$conversion{6}='6 _K';
$conversion{7}='7 _K';
$conversion{8}='8 _K';
$conversion{9}='9 _K';
$conversion{' '}='_s';


#parse terrain_file
open(TERRAIN, "<$terrain_file");
$countdef=0;
$max_len=0;
while($line=<TERRAIN>){

    if($line=~/^\#ifdef/){
#skip ifdef'd comments
	$countdef+=1;
        while(($countdef >0) && ($line=<TERRAIN>)){
	    if($line=~/^\#ifdef/){ 
		$countdef+=1;
	    }
	    if($line=~/^\#endif/){ 
		$countdef-=1;
	    }
	}
    }
    elsif($line=~/^\#/){}
    else{
#parse [terrain] blocks
	if($line=~/^\[terrain\]/){
	    $char='';
	    $string='';
	    $in=1;
	    while(($in>0) && ($line=<TERRAIN>)){
		$line=~s/\s+//;
		if($line=~/^char/){
		    $char=$line;
		    $char=~s/char//;
		    ($dummy,$char)=split('=',$char);
		    $char=~s/\"//g;
		    $char=~s/\s+//g;
		    $char=substr($char,0,1);
		}elsif($line=~/^string/){
		    $string=$line;
		    $string=~s/^string//;
		    ($dummy,$string)=split('=',$string);
		    $string=~s/\"//g;
		    $string=~s/\s+//g;
		}elsif($line=~/\[\/terrain\]/){
		    $in=0;
		    if((length($char)>0) && (length($string)>0)){
#			print "$char ---> $string\n";
			if(length($string)>$max_len){
			    $max_len=length($string);
			}
			$conversion{$char}=$string;
		    }
	        }
	    }
	}
    }
}
close(TERRAIN);

$width=$max_len+2;
open(MAP, "<$map_file");
@mfile=();
$map_only=1;
while($line=<MAP>){
    push(@mfile,$line);
    if($line=~/map_data/){
	$map_only=0;
    }

}

push(@mfile,"\n");

@map=();
close(MAP);

while($#mfile){
    $line=shift(@mfile);
    if($map_only || $line=~/map_data/){
	$cont=1;
#read map assumes map is more than 1 line long.
	if(!$map_only){
	    ($dummy,$line)=split('"',$line);
	}
	if((defined($line)) && (length($line))){push(@map,$line)};
#	 print "$line\n";
	 while(($cont) && ($#mfile)){
	     $line=shift(@mfile);
	     if($line=~/\"/){
		 $cont=0;
		($line,$dummy)=split('"',$line);
	     }
	     if(defined($line) && length($line)){push(@map,$line)};
       	 }

	if(! $map_only){ 
	    $line="map_data=\"\n";
	    push(@newfile,$line);
	}
	$y=0;
	 foreach(@map){
	     chomp;
	     if($_=~/,/){die "map file appears to be converted already\n";}
	     $line='';
	     for($x=0;$x!=length($_);$x++){
		 $hex='';
		 $char=substr($_,$x,1);
#		 print "$char";
		 $format="%${width}.${max_len}s";
		 if(defined($conversion{$char})){
		     $hex=sprintf($format,$conversion{$char});
		 }else{
		     $ord=ord($char);
		     die "error, unrecognized map character at ($x,$y):[$ord]$char";
#		     $hex=sprintf($format,$char);
		 }
		 if($hex=~/_K/){
		     #convert keeps according to adjacent hexes
		     $adj=get_adjacent($x,$y,@map);
#		     print "adjacent: $adj\n";
		     %hexcount=();
		     for($i=1;$i<length($adj);$i++){
			 #intentionally skipping 0 as it is original hex
			 $a=(substr($adj,$i,1));
			 $ca=$conversion{$a};
			 if(!defined($ca)){
			     $ord=ord($a);
			     print "error in adjacent hexes:\n";
			     print "($x,$y,$i)[$ord]:$a\n";
			 }
			 if($ca=~/^C/){ #this is a castle hex	
			     $hexcount{$ca}++;
			 }
		     }
		     $maxc=0;
		     $maxk="Ch";
		     foreach(keys(%hexcount)){
			 if($hexcount{$_}>$maxc){
			     $maxc=$hexcount{$_};
			     $maxk=$_;
			 }
#			 print "$_ $hexcount{$_}\n";
		     }
		     $maxk=~s/^C/K/;
		     $hex=~s/_K/$maxk/;
		 }
		 $line.=$hex;
		 if($x!=length($_)-1){$line.=','};
	     }
	     $line.="\n";
	     push(@newfile,$line);
#	     print "$line\n";
             $y++;
	 }
	if($map_only){
	    $line="\n";
	}else{
	    $line="\"\n";
	}
	push(@newfile,$line);
    }else{
	push(@newfile,$line);
    }
}

open(NEWMAP,">$new_map_file");
foreach(@newfile){
    print NEWMAP "$_";
}

File: data/tools/map_convert.py
#!/usr/bin/env python
# Script to convert from single char 1.2.x-style maps to multiple character
# 1.3.1-style maps.  This is a straight translation of the Perl created for
# regression-testing against the Perl version; the final version of the
# conversion logic will be folded into upconvert.

import sys, os, re

def printUsage():
    print "map_convert.pl terrain.cfg map_file.cfg [new_map_file.cfg]\n"

def get_adjacent(x, y, map):
    "Returns string of original location+adjacent locations on hex 1-char map"
    odd = (x) % 2
    adj = map[y][x];
    if x > 0:
	adj += map[y][x-1]
    if x < len(map[y])-1:
	adj += map[y][x+1]
    if y > 0:
	adj += map[y-1][x]
    if y < len(map)-1:
	adj += map[y+1][x]
    if x > 0 and y > 0 and not odd:
	adj += map[y-1][x-1]
    if x < len(map[y])-1 and y > 0 and not odd:
	adj += map[y-1][x+1];
    if x > 0 and y < len(map)-1 and odd:
	adj += map[y+1][x-1]
    if x < len(map[y])-1 and y < len(map)-1 and odd:
	adj += map[y+1][x+1]
    adj = adj.replace("\n", "").replace("\r", "")
    return adj

if len(sys.argv) < 3 or len(sys.argv) > 4:
    printUsage()
    sys.exit(1)

sys.argv.pop(0)
terrain_file = sys.argv.pop(0)
map_file = sys.argv.pop(0)
if not sys.argv:
    new_map_file = map_file;
    backup = map_file + ".bak";
    try:
        os.system("cp %s %s" % (map_file, backup))
    except:
        sys.stderr.write("map_convert: could not create backup file: %s\n" % backup)
        sys.exit(1)
else:
    new_map_file = sys.argv.pop(0)
    if os.path.exists(new_map_file):
        sys.stderr.write("map_convert: new map file already exists: %s\n" %new_map_file)
        sys.exit(1)

if not os.path.exists(terrain_file):
    sys.stderr.write("map_convert: can not read terrain file: %s\n"%terrain_file)
    sys.exit(1)

if not os.path.exists(map_file):
    sys.stderr.write("can not read map file: %s\n" % map_file)
    sys.exit(1)

conversion = {}
conversion["1"] = '1 _K'
conversion["2"] = '2 _K'
conversion["3"] = '3 _K'
conversion["4"] = '4 _K'
conversion["5"] = '5 _K'
conversion["6"] = '6 _K'
conversion["7"] = '7 _K'
conversion["8"] = '8 _K'
conversion["9"] = '9 _K'
conversion[" "] = '_s'

# Parse terrain_file
tfp = open(terrain_file)
countdef = 0
max_len = 0
while True:
    line = tfp.readline()
    if not line:
        break

    # Skip ifdef'd comments
    if line.startswith("#ifdef"):
	countdef +=1;
    if line.startswith("#endif"):
	countdef -=1;
    elif not line.startswith("#") and countdef == 0:
        # Parse [terrain] blocks
        if "[terrain]" in line:
	    char = ''
	    string = ''
	    inside = True
            while inside:
                line = tfp.readline()
                if not line:
                    break
		line = line.strip().replace(" ", "").replace("\t", "")
                if line.startswith("char"):
                    char = line[4:]
		    (dummy, char) = char.split('=');
		    char = char.replace('"', '')
                    if char:
                        char = char[0]
		elif line.startswith("string"):
                    string=line[6:]
		    (dummy, string) = string.split('=')
		    string = string.replace('"', '')
		elif line.startswith("[/terrain]"):
		    inside = False
                    if char and len(char) == 1 and string:
                        if len(string) > max_len:
			    max_len = len(string)
			conversion[char] = string;
tfp.close()
width = max_len+2

#keys = conversion.keys()
#keys.sort()
#for k in keys:
#    print "%s -> %s" % (k, conversion[k])

# This hairy regexp excludes map_data lines that contain {} file references,
# also lines that are empty or hold just one keep character (somewhat
# pathological, but not handling these will make the regression tests break).
mapdata = re.compile(r'map_data="[A-Za-z0-9\/|\\&_~?\[\]\']{2,}') 

mfile = []
map_only = not map_file.endswith(".cfg")
for line in open(map_file):
    mfile.append(line);
    if mapdata.search(line):
	map_only = False
#mfile.append("\n")

cont = False
outmap = []
newfile = []
lineno = baseline = 0
while mfile:
    line = mfile.pop(0)
    lineno += 1
    if map_only or mapdata.search(line):
        baseline = 0
	cont = True
	# Assumes map is more than 1 line long.
        if not map_only:
	    line = line.split('"')[1]
	if line:
            outmap.append(line)
        while cont and mfile:
            line = mfile.pop(0)
            lineno += 1
            if line and line[0] == '#':
                newfile.append(line)
                continue
            if '"' in line:
                cont = False
		line = line.split('"')[0]
                if line and not line.endswith("\n"):
                    line += "\n"
            if line:
                outmap.append(line)
	if not map_only: 
	    line="map_data=\"\n";
	    newfile.append(line)
	y = 0
        for oldline in outmap:
            x = 0
            if "," in oldline:
                sys.stderr.write("mapconvert map file %s appears "
                                 "to be converted already\n" % map_file)
                sys.exit(1)
            line = ''
            for char in oldline:
                ohex = ''
                format = "%%%d.%ds" % (width, max_len)
                if char in ('\n', '\r'):
                    ohex += char
                elif char in conversion:
                    ohex = format % conversion[char] + ','
                else:
                    sys.stderr.write('map_convert: "%s", line %d: unrecognized character %s (%d) at (%d, %d)\n' % \
                                     (map_file, baseline+y+1, `char`, ord(char), x, y))
                    # ohex = format % char
                    sys.exit(1)
                if "_K" in ohex:
                    # Convert keeps according to adjacent hexes
                    adj = get_adjacent(x,y, outmap)
                    # print "adjacent: %s" % adj
                    hexcount = {}
                    for i in range(1, len(adj)):
                        # Intentionally skipping 0 as it is original hex
                        a = adj[i];
                        if not a in conversion:
                            sys.stderr.write('map_convert: "%s", line %d: error in adjacent hexes at (%d, %d).\n' % \
                                     (map_file, baseline+y+1, x, y))
                            sys.exit(1)
                        ca = conversion[a]
                        if ca.startswith("C"): #this is a castle hex	
			     hexcount[ca] = hexcount.get(ca, 0) + 1
                    maxc = 0;
                    maxk = "Ch";
                    # Next line is a hack to make this code pass regression
                    # testing against the Perl original. Without the sort, when
                    # there are two terrain types that occur in equal numbers
                    # greater than any others, which one gets picked will be
                    # randomly dependent on Python's dictionary hash function.
                    sorted = hexcount.keys()
                    sorted.sort()
                    for k in sorted:
                        if hexcount[k] > maxc:
                            maxc = hexcount[k]
                            maxk = k
                    #print "Dominated by %s" % maxk
                    maxk = re.sub("^C", "K", maxk)
                    ohex = ohex.replace("_K", maxk)
                line += ohex
                x += 1
            # We've accumulated a translated map line
            newfile.append(line.replace(",\n", "\n"))
            y += 1
        # All lines of the map are processed, add the appropriate trailer
        if map_only:
	    line="\n"
	else:
	    line="\"\n"
	newfile.append(line)
    else:
	newfile.append(line)

ofp = open(new_map_file, "w");
ofp.writelines(newfile)
ofp.close()

# map_convert ends here.


